File: app/src/main/java/com/tunjid/fingergestures/viewholders/WallpaperTriggerViewHolder.java
/*
 * Copyright (c) 2017, 2018, 2019 Adetunji Dahunsi.
 *
 *     This program is free software: you can redistribute it and/or modify
 *     it under the terms of the GNU General Public License as published by
 *     the Free Software Foundation, either version 3 of the License, or
 *     (at your option) any later version.
 *
 *     This program is distributed in the hope that it will be useful,
 *     but WITHOUT ANY WARRANTY; without even the implied warranty of
 *     MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 *     GNU General Public License for more details.
 *
 *     You should have received a copy of the GNU General Public License
 *     along with this program.  If not, see <https://www.gnu.org/licenses/>.
 */

package com.tunjid.fingergestures.viewholders;

import android.app.TimePickerDialog;
import android.graphics.drawable.Drawable;
import androidx.annotation.DrawableRes;
import androidx.core.content.ContextCompat;
import android.view.View;
import android.widget.TextView;

import com.tunjid.fingergestures.App;
import com.tunjid.fingergestures.R;
import com.tunjid.fingergestures.BackgroundManager;
import com.tunjid.fingergestures.adapters.AppAdapter;

import java.text.DateFormat;
import java.text.SimpleDateFormat;
import java.util.Calendar;
import java.util.Locale;

import static com.tunjid.fingergestures.BackgroundManager.NIGHT_WALLPAPER_PICK_CODE;
import static com.tunjid.fingergestures.BackgroundManager.DAY_WALLPAPER_PICK_CODE;
import static java.util.Calendar.HOUR_OF_DAY;
import static java.util.Calendar.MINUTE;

public class WallpaperTriggerViewHolder extends AppViewHolder {

    private static final DateFormat dateFormat = new SimpleDateFormat("h:mm a", Locale.US);

    private TextView start;
    private TextView end;
    private BackgroundManager backgroundManager;

    public WallpaperTriggerViewHolder(View itemView, AppAdapter.AppAdapterListener appAdapterListener) {
        super(itemView, appAdapterListener);

        start = itemView.findViewById(R.id.start);
        end = itemView.findViewById(R.id.end);
        backgroundManager = BackgroundManager.getInstance();

        start.setOnClickListener(view -> showTimePicker(backgroundManager.getMainWallpaperCalendar(DAY_WALLPAPER_PICK_CODE),
                (dialog, hour, minute) -> {
                    backgroundManager.setWallpaperChangeTime(DAY_WALLPAPER_PICK_CODE, hour, minute);
                    bind();
                }));

        end.setOnClickListener(view -> showTimePicker(backgroundManager.getMainWallpaperCalendar(NIGHT_WALLPAPER_PICK_CODE),
                (dialog, hour, minute) -> {
                    backgroundManager.setWallpaperChangeTime(NIGHT_WALLPAPER_PICK_CODE, hour, minute);
                    bind();
                }));

        itemView.findViewById(R.id.cancel_auto_wallpaper).setOnClickListener(view -> {
            backgroundManager.cancelAutoWallpaper();
            bind();
        });
    }

    @Override
    public void bind() {
        super.bind();
        start.setCompoundDrawablesWithIntrinsicBounds(null, getTopDrawable(DAY_WALLPAPER_PICK_CODE), null, null);
        end.setCompoundDrawablesWithIntrinsicBounds(null, getTopDrawable(NIGHT_WALLPAPER_PICK_CODE), null, null);
        start.setText(getCalendarString(backgroundManager.getMainWallpaperCalendar(DAY_WALLPAPER_PICK_CODE)));
        end.setText(getCalendarString(backgroundManager.getMainWallpaperCalendar(NIGHT_WALLPAPER_PICK_CODE)));
        start.setTextColor(getTextColor(DAY_WALLPAPER_PICK_CODE));
        end.setTextColor(getTextColor(NIGHT_WALLPAPER_PICK_CODE));
    }

    private void showTimePicker(Calendar calendar, TimePickerDialog.OnTimeSetListener listener) {
        if (App.hasStoragePermission())
            new TimePickerDialog(itemView.getContext(), listener, calendar.get(HOUR_OF_DAY), calendar.get(MINUTE), false).show();
        else adapterListener.showSnackbar(R.string.enable_storage_settings);
    }

    private String getCalendarString(Calendar calendar) {
        return dateFormat.format(calendar.getTime());
    }

    private int getTextColor(@BackgroundManager.WallpaperSelection int selection) {
        return ContextCompat.getColor(itemView.getContext(), backgroundManager.willChangeWallpaper(selection)
                ? R.color.toggle_text
                : R.color.dark_grey);
    }

    private Drawable getTopDrawable(@BackgroundManager.WallpaperSelection int selection) {
        @DrawableRes int drawableRes = selection == DAY_WALLPAPER_PICK_CODE ? R.drawable.ic_day_24dp : R.drawable.ic_night_24dp;
        return backgroundManager.tint(drawableRes, getTextColor(selection));
    }
}

File: app/src/main/java/com/tunjid/fingergestures/viewholders/WallpaperTriggerViewHolder.kt
/*
 * Copyright (c) 2017, 2018, 2019 Adetunji Dahunsi.
 *
 *     This program is free software: you can redistribute it and/or modify
 *     it under the terms of the GNU General Public License as published by
 *     the Free Software Foundation, either version 3 of the License, or
 *     (at your option) any later version.
 *
 *     This program is distributed in the hope that it will be useful,
 *     but WITHOUT ANY WARRANTY; without even the implied warranty of
 *     MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 *     GNU General Public License for more details.
 *
 *     You should have received a copy of the GNU General Public License
 *     along with this program.  If not, see <https://www.gnu.org/licenses/>.
 */

package com.tunjid.fingergestures.viewholders

import android.app.TimePickerDialog
import android.graphics.drawable.Drawable
import android.view.ViewGroup
import com.tunjid.androidx.core.content.colorAt
import com.tunjid.androidx.core.content.drawableAt
import com.tunjid.androidx.core.graphics.drawable.withTint
import com.tunjid.androidx.recyclerview.viewbinding.BindingViewHolder
import com.tunjid.androidx.recyclerview.viewbinding.viewHolderDelegate
import com.tunjid.androidx.recyclerview.viewbinding.viewHolderFrom
import com.tunjid.fingergestures.models.uiState
import com.tunjid.fingergestures.models.updatePartial
import com.tunjid.fingergestures.App
import com.tunjid.fingergestures.R
import com.tunjid.fingergestures.WallpaperSelection
import com.tunjid.fingergestures.adapters.Item
import com.tunjid.fingergestures.databinding.ViewholderWallpaperTriggerBinding
import java.text.SimpleDateFormat
import java.util.*
import java.util.Calendar.HOUR_OF_DAY
import java.util.Calendar.MINUTE

private var BindingViewHolder<ViewholderWallpaperTriggerBinding>.item by viewHolderDelegate<Item.WallpaperTrigger>()

fun ViewGroup.wallpaperTrigger() = viewHolderFrom(ViewholderWallpaperTriggerBinding::inflate).apply {
    binding.start.setOnClickListener { selectTime(WallpaperSelection.Day) }
    binding.end.setOnClickListener { selectTime(WallpaperSelection.Night) }
    binding.cancelAutoWallpaper.setOnClickListener { item.cancelAutoWallpaper() }
}

fun BindingViewHolder<ViewholderWallpaperTriggerBinding>.bind(item: Item.WallpaperTrigger) = binding.run {
    this@bind.item = item
    start.setCompoundDrawablesWithIntrinsicBounds(null, getTopDrawable(WallpaperSelection.Day), null, null)
    end.setCompoundDrawablesWithIntrinsicBounds(null, getTopDrawable(WallpaperSelection.Night), null, null)
    start.text = getCalendarString(item.dayStatus.calendar)
    end.text = getCalendarString(item.nightStatus.calendar)
    start.setTextColor(getTextColor(WallpaperSelection.Day))
    end.setTextColor(getTextColor(WallpaperSelection.Night))
}

private fun BindingViewHolder<ViewholderWallpaperTriggerBinding>.showTimePicker(calendar: Calendar, onTimeSetListener: TimePickerDialog.OnTimeSetListener) {
    if (App.hasStoragePermission)
        TimePickerDialog(itemView.context, onTimeSetListener, calendar.get(HOUR_OF_DAY), calendar.get(MINUTE), false).show()
    else itemView::uiState.updatePartial { copy(snackbarText = itemView.context.getString(R.string.enable_storage_settings)) }
}

private fun getCalendarString(calendar: Calendar): String = dateFormat.format(calendar.time)

private fun BindingViewHolder<ViewholderWallpaperTriggerBinding>.getTextColor(selection: WallpaperSelection): Int =
    itemView.context.colorAt(when {
        when (selection) {
            WallpaperSelection.Day -> item.dayStatus.willChange
            WallpaperSelection.Night -> item.nightStatus.willChange
        } -> R.color.toggle_text
        else -> R.color.dark_grey
    })

private fun BindingViewHolder<ViewholderWallpaperTriggerBinding>.getTopDrawable(selection: WallpaperSelection): Drawable? =
    itemView.context.drawableAt(when (selection) {
        WallpaperSelection.Day -> R.drawable.ic_day_24dp
        WallpaperSelection.Night -> R.drawable.ic_night_24dp
    })?.withTint(getTextColor(selection))

private fun BindingViewHolder<ViewholderWallpaperTriggerBinding>.selectTime(selection: WallpaperSelection) =
    showTimePicker(when (selection) {
        WallpaperSelection.Day -> item.dayStatus.calendar
        WallpaperSelection.Night -> item.nightStatus.calendar
    }) { _, hour, minute -> item.selectTime(selection, hour, minute) }

private val dateFormat = SimpleDateFormat("h:mm a", Locale.US)
